File: bending-api/src/main/java/pw/dotdash/bending/api/ray/AirRaycast.java
package pw.dotdash.bending.api.ray;

import com.flowpowered.math.vector.Vector3d;
import org.spongepowered.api.block.BlockTypes;
import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.entity.Entity;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.world.Location;
import org.spongepowered.api.world.World;
import pw.dotdash.bending.api.effect.EffectService;
import pw.dotdash.bending.api.util.BlockTypeUtil;
import pw.dotdash.bending.api.util.VectorUtil;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * A collection of utilities primarily used by air abilities' raycasts.
 */
public final class AirRaycast {

    public static boolean extinguishFlames(Location<World> test) {
        checkNotNull(test, "test");

        if (test.getBlockType() != BlockTypes.FIRE) {
            return false;
        }

        test.setBlockType(BlockTypes.AIR);
        test.getExtent().spawnParticles(EffectService.getInstance().getExtinguishEffect(), test.getPosition());

        return true;
    }

    public static boolean coolLava(Location<World> test) {
        checkNotNull(test, "test");

        if (test.getBlockType() != BlockTypes.LAVA && test.getBlockType() != BlockTypes.FLOWING_LAVA) {
            return false;
        }

        if (test.getBlockType() == BlockTypes.FLOWING_LAVA) {
            test.setBlockType(BlockTypes.AIR);
        } else if (test.get(Keys.FLUID_LEVEL).get() == 0) {
            test.setBlockType(BlockTypes.OBSIDIAN);
        } else {
            test.setBlockType(BlockTypes.COBBLESTONE);
        }

        return true;
    }

    public static boolean toggleDoor(Location<World> test) {
        checkNotNull(test, "test");

        if (!BlockTypeUtil.DOORS.contains(test.getBlockType())) {
            return false;
        }

        test.offer(Keys.OPEN, !test.getOrElse(Keys.OPEN, false));

        return true;
    }

    public static boolean toggleLever(Location<World> test) {
        checkNotNull(test, "test");

        if (test.getBlockType() != BlockTypes.LEVER) {
            return false;
        }

        test.offer(Keys.POWERED, !test.getOrElse(Keys.POWERED, false));

        return true;
    }

    public static boolean pushEntity(Raycast ray, Player source, Entity target,
                              boolean canPushSelf, double knockbackSelf, double knockbackOther) {
        checkNotNull(ray, "ray");
        checkNotNull(source, "source");
        checkNotNull(target, "target");

        final boolean isSelf = source.getUniqueId().equals(target.getUniqueId());
        double knockback = knockbackOther;

        if (isSelf) {
            if (!canPushSelf) {
                // Ignore us.
                return false;
            }

            knockback = knockbackSelf;
        }

        knockback *= (1 - target.getLocation().getPosition().distance(ray.getOrigin().getPosition()) / (2 * ray.getRange()));

        if (BlockTypeUtil.isSolid(target.getLocation().add(0.0, -0.5, 0.0).getBlockType())) {
            knockback *= 0.85;
        }

        final double max = ray.getSpeed() / ray.getSpeedFactor();

        Vector3d push = ray.getDirection();
        if (Math.abs(push.getY()) > max && !isSelf) {
            push = VectorUtil.setY(push, push.getY() < 0 ? -max : max);
        }

        push = push.normalize().mul(knockback);

        if (Math.abs(target.getVelocity().dot(push)) > knockback
                && VectorUtil.getAngleBetween(target.getVelocity(), push) > Math.PI / 3) {
            push = push.normalize().add(target.getVelocity()).mul(knockback);
        }

        target.setVelocity(push.min(4.0, 4.0, 4.0).max(-4.0, -4.0, -4.0));
        return true;
    }

    private AirRaycast() {
        throw new AssertionError("Don't instantiate me!");
    }
}

File: src/main/kotlin/arvenwood/bending/plugin/raycast/AirRaycast.kt
package arvenwood.bending.plugin.projectile

import arvenwood.bending.api.util.*
import arvenwood.bending.plugin.ability.air.AirConstants
import org.spongepowered.api.block.BlockTypes
import org.spongepowered.api.data.key.Keys
import org.spongepowered.api.effect.sound.SoundType
import org.spongepowered.api.effect.sound.SoundTypes
import org.spongepowered.api.entity.Entity
import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.world.Location
import org.spongepowered.api.world.World
import kotlin.math.absoluteValue

object AirRaycast {

    fun extinguishFlames(test: Location<World>): Boolean {
        if (test.blockType != BlockTypes.FIRE) return false

        test.blockType = BlockTypes.AIR
        test.spawnParticles(AirConstants.EXTINGUISH_EFFECT)

        return true
    }

    fun coolLava(test: Location<World>): Boolean {
        if (test.blockType != BlockTypes.LAVA && test.blockType != BlockTypes.FLOWING_LAVA) return false

        when {
            test.blockType == BlockTypes.FLOWING_LAVA -> test.blockType = BlockTypes.AIR
            test.get(Keys.FLUID_LEVEL).get() == 0 -> test.blockType = BlockTypes.OBSIDIAN
            else -> test.blockType = BlockTypes.COBBLESTONE
        }

        return true
    }

    fun toggleDoor(test: Location<World>): Boolean {
        if (test.blockType !in AirConstants.DOORS) return false

        // Open/Close doors.
        val open: Boolean = test.get(Keys.OPEN).orElse(false)
        test.offer(Keys.OPEN, !open)
        test.extent.playSound(doorSound(!open), test.position, 0.5, 0.0)

        return true
    }

    private fun doorSound(open: Boolean): SoundType =
        if (open) SoundTypes.BLOCK_WOODEN_DOOR_OPEN else SoundTypes.BLOCK_WOODEN_DOOR_CLOSE

    fun toggleLever(test: Location<World>): Boolean {
        if (test.blockType != BlockTypes.LEVER) return false

        // Flip switches.
        test.offer(Keys.POWERED, test.get(Keys.POWERED).orElse(false))
        test.extent.playSound(SoundTypes.BLOCK_LEVER_CLICK, test.position, 0.5, 0.0)

        return true
    }

    fun Raycast.pushEntity(source: Player, target: Entity, canPushSelf: Boolean, pushFactorSelf: Double, pushFactorOther: Double): Boolean {
        val isSelf: Boolean = source.uniqueId == target.uniqueId
        var knockback: Double = pushFactorOther

        if (isSelf) {
            if (!canPushSelf) {
                // Ignore us.
                return false
            }

            knockback = pushFactorSelf
        }

        val max: Double = this.speed / this.speedFactor

        var push = this.direction
        if (push.y.absoluteValue > max && !isSelf) {
            push = push.withY(if (push.y < 0) -max else max)
        }

        knockback *= (1 - target.position.distance(this.origin.position) / (2 * this.range))

        if (target.location.add(0.0, -0.5, 0.0).blockType.isSolid()) {
            knockback *= 0.85
        }

        push = push.normalize() * knockback
        if (target.velocity.dot(push).absoluteValue > knockback && target.velocity.angle(push) > Math.PI / 3) {
            // Increase the velocity in their current direction.
            push = push.normalize().add(target.velocity) * knockback
        }

        // Toss the target.
        target.velocity = push.min(4.0, 4.0, 4.0).max(-4.0, -4.0, -4.0)
        return true
    }
}